### PR TITLE
Change client behavior on 503 from retry to failover

### PR DESCRIPTION
**Pull request by gregakinman**
@uschi2000 

Fixes https://github.com/palantir/http-remoting/issues/678

### CODE REVIEW

**Comment by uschi2000** [comment]
this looks good. i'll have to run for dinner now, but will take a more in-depth look later.

**Comment by gregakinman** [comment]
Ugh looks like the build failed :( I’ll be able to take a look later tonight

File: jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/JaxRsClientFailoverTest.java [hunk 1]
```diff
@@ -118,10 +118,23 @@ public void testConnectionError_performsFailoverOnDnsFailure() throws Exception
 
     @Test
     public void testQosError_performsFailover() throws Exception {
-        server1.enqueue(new MockResponse().setResponseCode(503));
+        server1.enqueue(new MockResponse().setResponseCode(UNAVAILABLE_RESPONSE_CODE));
```
Comment (uschi2000) [inline jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/JaxRsClientFailoverTest.java]: think I prefer the 503 here

Comment (uschi2000) [inline jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/JaxRsClientFailoverTest.java]: think I prefer the 503 here

Comment (uschi2000) [reply jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/JaxRsClientFailoverTest.java]: can you add a similar test for okhttp-clients?

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java [hunk 2]
```diff
@@ -239,48 +239,44 @@ public Void visit(QosException.Throttle exception) {
...
-                    callback.onFailure(call,
-                            new IOException("Failed to complete the request due to a "
-                                    + "server-side QoS condition: 503", exception));
-                } else {
-                    log.debug("Rescheduling call after backoff",
-                            SafeArg.of("backoffMillis", backoff.get().toMillis()));
-                    scheduleExecution(callback, backoff.get());
-                }
-                return null;
+                return redirectToUrl(callback, call,
```
Comment (uschi2000) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java]: I'd like to retry another host *with backoff*

**Review by uschi2000 — CHANGES_REQUESTED** [review]
should perform backoff on 503

File: jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/JaxRsClientFailoverTest.java (hunk 1 above)
Comment (gregakinman) [reply jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/JaxRsClientFailoverTest.java]: sure

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java (hunk 2 above)
Comment (gregakinman) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java]: Are you thinking of the single-node case here? I thought about this and ended up going this way just to index on the multi-node case - if we get a 503 from one node, we have no data on the health of the other nodes, thus we can safely assume (as we always do when we are issuing the first try of a request) that any other node is healthy right now, and we can try immediately with no backoff. But in the single-node case we'd definitely want to back off. But the logic of how many URLs there are shouldn't be present at this level... and I guess you could run into the case where _all_ nodes are unavailable due to load, and the client is getting bounced around, in which case it still makes sense to back off. And backing off the first time or two probably isn't the worst penalty to pay. Okay.

Comment (uschi2000) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java]: exactly

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java [hunk 3]
```diff
@@ -212,10 +213,10 @@ public void onResponse(Call call, Response response) throws IOException {
     }
 
     @SuppressWarnings("FutureReturnValueIgnored")
-    private void scheduleExecution(Callback callback, Duration backoff) {
+    private void scheduleExecution(Callback callback, Duration backoff, Runnable execution) {
```
Comment (uschi2000) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java]: callback is now unused?

Comment (uschi2000) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java]: prefer to order `(execution, backoff)`

Comment (gregakinman) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java]: whoops, fixed

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java [hunk 4]
```diff
@@ -274,13 +263,38 @@ public Void visit(QosException.Unavailable exception) {
...
                             SafeArg.of("backoffMillis", backoff.get().toMillis()));
-                    scheduleExecution(callback, backoff.get());
+                    redirectToUrl(
+                            callback,
+                            call,
+                            // Redirect to the "next" URL, whichever that may be, after backing off.
+                            () -> urls.redirectToNext(request().url()),
+                            "Failed to determine valid redirect URL for base URLs " + urls.getBaseUrls(),
+                            request -> scheduleExecution(() -> client.newCallWithMutableState(request,
+                                    backoffStrategy, maxNumRelocations - 1).enqueue(callback), backoff.get()));
```
Comment (uschi2000) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java]: don't think we need to decrease maxNumRelocations in this case. backoff already limits the number of retries for 503s.

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java [hunk 5]
```diff
@@ -274,13 +263,38 @@ public Void visit(QosException.Unavailable exception) {
...
                 }
                 return null;
             }
         };
     }
 
+    private void redirectToUrl(Callback callback, Call call, Supplier<Optional<HttpUrl>> optionalUrlSupplier,
```
Comment (uschi2000) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java]: think this construction is pretty hard to grok, maybe be more readable with a little bit of code duplication in the visit() methods. summarizing:
- RetryOther should fail if maxNumRetries <= 0 or no URL can be determined
- Unavailable should fail if backoff is empty or no URL can be determined

File: okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/OkHttpClientsTest.java [hunk 6]
```diff
@@ -401,18 +402,17 @@ public void handlesQos_canRetryLaterAndThenRedirect() throws Exception {
...
 
     @Test
-    public void handlesQos_memorizedCurrentUrlBetweenCalls() throws Exception {
-        // First hits server,then 308 redirects to server2, then retries, waits on 503, then retries server2 again.
-        server.enqueue(new MockResponse().setResponseCode(308).addHeader(HttpHeaders.LOCATION, url2));
-        server2.enqueue(new MockResponse().setResponseCode(503));
-        server2.enqueue(new MockResponse().setResponseCode(200).setBody("foo"));
+    public void handlesQos_failsOverToAnotherUrl() throws Exception {
+        server.enqueue(new MockResponse().setResponseCode(503));
+        server.enqueue(new MockResponse().setBody("foo"));
```
Comment (uschi2000) [inline okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/OkHttpClientsTest.java]: remove this line

File: okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/OkHttpClientsTest.java [hunk 7]
```diff
@@ -401,18 +402,17 @@ public void handlesQos_canRetryLaterAndThenRedirect() throws Exception {
     }
 
     @Test
-    public void handlesQos_memorizedCurrentUrlBetweenCalls() throws Exception {
-        // First hits server,then 308 redirects to server2, then retries, waits on 503, then retries server2 again.
-        server.enqueue(new MockResponse().setResponseCode(308).addHeader(HttpHeaders.LOCATION, url2));
-        server2.enqueue(new MockResponse().setResponseCode(503));
-        server2.enqueue(new MockResponse().setResponseCode(200).setBody("foo"));
+    public void handlesQos_failsOverToAnotherUrl() throws Exception {
```
Comment (uschi2000) [inline okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/OkHttpClientsTest.java]: handlesQos_503FailsOverToAnotherUrl

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java (hunk 5 above)
Comment (gregakinman) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java]: yeah realized i was possibly going overboard with the DRY there

**Review by uschi2000 — APPROVED** [review]

**Comment by uschi2000** [comment]
thanks, @gregakinman

**Comment by gregakinman** [comment]
np! would like to cut 3.19.0 with this too if you weren't planning to already

**Pull request merged**